// @progress/kendo-react-spreadsheet/List.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import { IconWrap as m } from "@progress/kendo-react-common";
import { formulaFxIcon as c } from "@progress/kendo-svg-icons";
const n = e.forwardRef((a, r) => {
  const l = e.useRef(null), s = e.useRef(null);
  return e.useImperativeHandle(l, () => ({
    element: s.current,
    props: a
  })), e.useImperativeHandle(r, () => l.current), /* @__PURE__ */ e.createElement("div", { className: "k-list k-list-md" }, /* @__PURE__ */ e.createElement("div", { className: "k-list-content" }, /* @__PURE__ */ e.createElement(
    "ul",
    {
      ref: s,
      className: "k-spreadsheet-formula-list k-list-ul",
      onMouseDown: (t) => t.preventDefault(),
      style: { maxHeight: 280 }
    },
    a.data.map(
      (t) => /* @__PURE__ */ e.createElement("li", { key: t.text, className: "k-list-item", onClick: () => a.onItemClick(t.value) }, /* @__PURE__ */ e.createElement(m, { name: "formula-fx", icon: c }), /* @__PURE__ */ e.createElement("span", { className: "k-list-item-text" }, t.text))
    )
  )));
});
n.displayName = "List";
n.propTypes = {};
export {
  n as List
};
