// @progress/kendo-react-spreadsheet/tools/cleanFormat.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import { clearCssIcon as a } from "@progress/kendo-svg-icons";
import { Button as c } from "@progress/kendo-react-buttons";
import { useLocalization as m } from "@progress/kendo-react-intl";
import { keys as o, messages as s } from "../messages.mjs";
const p = (n) => {
  const { spreadsheetRef: t } = n, r = e.useCallback(() => {
    t.current && t.current.executeCommand({ command: "PropertyCleanCommand" });
  }, []);
  return /* @__PURE__ */ e.createElement(
    c,
    {
      type: "button",
      icon: "clean-css",
      svgIcon: a,
      fillMode: "flat",
      onClick: r,
      title: m().toLanguageString(o.cleanFormatting, s[o.cleanFormatting])
    }
  );
};
export {
  p as CleanFormatting
};
