// @progress/kendo-react-spreadsheet/tools/gridLines.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as o from "react";
import { Button as c } from "@progress/kendo-react-buttons";
import { bordersNoneIcon as m } from "@progress/kendo-svg-icons";
import { useLocalization as l } from "@progress/kendo-react-intl";
import { keys as n, messages as d } from "../messages.mjs";
const u = (r) => {
  const { value: i, spreadsheetRef: e } = r, s = o.useCallback(() => {
    if (e.current) {
      const t = e.current.activeSheet();
      if (t) {
        const a = {
          command: "GridLinesChangeCommand",
          options: { property: "gridLines", value: !t.showGridLines() }
        };
        e.current.executeCommand(a);
      }
    }
  }, []);
  return /* @__PURE__ */ o.createElement(
    c,
    {
      type: "button",
      togglable: !0,
      icon: "borders-none",
      svgIcon: m,
      fillMode: "flat",
      title: l().toLanguageString(n.gridLines, d[n.gridLines]),
      onClick: s,
      selected: i
    }
  );
};
u.displayName = "GridLines";
export {
  u as GridLines
};
