// @progress/kendo-react-spreadsheet/tools/textWrap.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as o from "react";
import { Button as p } from "@progress/kendo-react-buttons";
import { textWrapIcon as i } from "@progress/kendo-svg-icons";
import { useLocalization as l } from "@progress/kendo-react-intl";
import { keys as r, messages as m } from "../messages.mjs";
const u = (a) => {
  const { value: n, spreadsheetRef: e } = a, c = o.useCallback(() => {
    if (e.current) {
      const t = e.current.activeSheet();
      if (t) {
        const s = {
          command: "TextWrapCommand",
          options: { property: "wrap", value: !t.range(t.activeCell()).wrap() }
        };
        e.current.executeCommand(s);
      }
    }
  }, []);
  return /* @__PURE__ */ o.createElement(
    p,
    {
      type: "button",
      togglable: !0,
      icon: "text-wrap",
      svgIcon: i,
      fillMode: "flat",
      title: l().toLanguageString(r.textWrap, m[r.textWrap]),
      onClick: c,
      selected: n
    }
  );
};
u.displayName = "TextWrap";
export {
  u as TextWrap
};
