// @progress/kendo-react-dropdowns/common/List.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import Q from "./ListItem.mjs";
import T from "./ListGroupItem.mjs";
import { areSame as R, getItemValue as b } from "./utils.mjs";
import { useLocalization as W } from "@progress/kendo-react-intl";
import { nodata as k, messages as X } from "../messages/index.mjs";
import { useUnstyled as Y, classNames as g, uDropDownsBase as p } from "@progress/kendo-react-common";
const le = (e) => {
  const B = W(), y = Y(), c = y && y.uDropDownsBase, {
    id: L,
    show: z,
    wrapperCssClass: M,
    wrapperStyle: $,
    listStyle: x,
    listRef: A,
    wrapperRef: F,
    listClassName: G,
    ariaSetSize: V
  } = e, D = (() => {
    const {
      textField: n,
      valueField: l,
      groupField: u,
      groupMode: s,
      isMultiColumn: U,
      optionsGuid: I,
      skip: S = 0,
      virtual: h,
      focusedIndex: C,
      hasDuplicates: j,
      highlightSelected: q = !0,
      value: i,
      data: E,
      itemRender: J,
      groupHeaderItemRender: K
    } = e, w = Array.isArray(i);
    let N = 0;
    return E.map((o, r) => {
      const m = S + r, f = S + r + N, O = j ? i ? C === r : !1 : o.disabled ? !1 : q && (!w && R(o, i, l) || w && i.findIndex((P) => R(P, o, l)) !== -1);
      let a, d, v;
      return r > 0 && u !== void 0 && (d = b(o, u), v = b(E[r - 1], u), d && v && d !== v && (a = d)), a !== void 0 && s === "modern" && (N += 1), [
        a !== void 0 && s === "modern" && /* @__PURE__ */ t.createElement(
          T,
          {
            id: `option-${I}-${f}`,
            virtual: h,
            key: m + "-group-item",
            group: a,
            isMultiColumn: U,
            render: K
          }
        ),
        /* @__PURE__ */ t.createElement(
          Q,
          {
            id: `option-${I}-${a !== void 0 && s === "modern" ? f + 1 : f}`,
            virtual: h,
            dataItem: o,
            groupMode: s,
            selected: O,
            focused: C === r,
            index: m,
            key: m,
            onClick: e.onClick,
            textField: n,
            group: a,
            render: J,
            disabled: o.disabled
          }
        )
      ];
    });
  })(), H = () => {
    const n = e.noDataRender, l = /* @__PURE__ */ t.createElement("div", { className: g(p.noData({ c })) }, /* @__PURE__ */ t.createElement("div", null, B.toLanguageString(k, X[k])));
    return n ? n.call(void 0, l) : l;
  };
  return D.length ? /* @__PURE__ */ t.createElement(
    "div",
    {
      className: M,
      style: $,
      ref: F,
      onMouseDown: e.onMouseDown,
      onBlur: e.onBlur,
      onScroll: e.onScroll,
      unselectable: "on"
    },
    /* @__PURE__ */ t.createElement(
      "ul",
      {
        id: L,
        role: "listbox",
        "aria-hidden": z ? void 0 : !0,
        "aria-setsize": V,
        className: G || g(p.ul({ c })),
        ref: A,
        style: x
      },
      D
    ),
    e.scroller && /* @__PURE__ */ t.createElement("div", { className: g(p.heightContainer({ c })) }, e.scroller)
  ) : H();
};
export {
  le as default
};
