// @progress/kendo-react-common/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "./tree-utils/itemIdUtils.mjs";
import { ZIndexContext as t, useZIndexContext as p } from "./contexts/ZIndexContext.mjs";
import { BrowserSupportService as a } from "./browser-support.service.mjs";
import { canUseDOM as n } from "./canUseDOM.mjs";
import { canUseRef as x } from "./canUseRef.mjs";
import { classNames as l, strippedClassNames as u } from "./classNames.mjs";
import { clone as c, cloneArray as I, cloneDate as C, cloneObject as h, cloneValue as b } from "./clone.mjs";
import { applyTrappedFocus as E, disableNavigatableContainer as D, enableNavigatableContainer as P, firstFocusableChild as A, focusFirstFocusableChild as S, focusLastFocusableChild as T, focusableChildren as B, keepFocusInContainer as v, lastFocusableChild as F } from "./trappedFocus.mjs";
import { FormComponent as L } from "./FormComponent.mjs";
import { getTabIndex as w } from "./getTabIndex.mjs";
import { getter as k } from "./getter.mjs";
import { guid as U } from "./guid.mjs";
import { KEYS as W, Keys as _ } from "./keys.mjs";
import { noop as H } from "./noop.mjs";
import { getScrollbarWidth as j, setScrollbarWidth as z } from "./scrollbarWidth.mjs";
import { setter as Z } from "./setter.mjs";
import { FIELD_REGEX as Y, FOCUSABLE_ELEMENTS as q, FOCUSABLE_ELEMENTS_BASE as J, TABBABLE_ELEMENTS as Q } from "./constants/main.mjs";
import { extendDataItem as ee, getItemPath as oe, mapTree as re, mapTreeItem as te } from "./treeDataOperations.mjs";
import { shouldShowValidationUI as se, validatePackage as ae } from "./validate-package.mjs";
import { hasRelativeStackingContext as ne } from "./hasRelativeStackingContext.mjs";
import { WatermarkOverlay as xe } from "./watermark/WatermarkOverlay.mjs";
import { dispatchEvent as le } from "./events/dispatchEvent.mjs";
import { AsyncFocusBlur as de } from "./hocs/AsyncFocusBlur.mjs";
import { createPropsContext as Ie, withPropsContext as Ce } from "./hocs/withPropsContext.mjs";
import { withIdHOC as be } from "./hocs/use-id-hoc.mjs";
import { withUnstyledHOC as Ee } from "./hocs/use-unstyled-hoc.mjs";
import { useDir as Pe } from "./hooks/useDir.mjs";
import { useAsyncFocusBlur as Se } from "./hooks/useAsyncFocusBlur.mjs";
import { useRtl as Be } from "./hooks/useRtl.mjs";
import { useMouse as Fe } from "./hooks/useMouse.mjs";
import { useCustomComponent as Le } from "./hooks/useCustomComponent.mjs";
import { usePropsContext as we } from "./hooks/usePropsContext.mjs";
import { COLLECTION_ACTION as ke, useCollection as Oe } from "./hooks/useCollection.mjs";
import { useDocument as Re } from "./hooks/useDocument.mjs";
import { useWindow as _e } from "./hooks/useWindow.mjs";
import { useId as He } from "./hooks/use-id.mjs";
import { useIsomorphicLayoutEffect as je } from "./hooks/use-isomorphic-layout-effect.mjs";
import { Icon as Ke } from "./icons/Icon.mjs";
import { SvgIcon as Xe } from "./icons/SvgIcon.mjs";
import { IconWrap as qe } from "./icons/IconWrap.mjs";
import { IconsContext as Qe } from "./icons/IconsContext.mjs";
import { svgIconPropType as eo, toIconClass as oo, toIconName as ro } from "./icons/utils.mjs";
import { Typography as po } from "./typography/Typography.mjs";
import { DragAndDrop as ao } from "./drag-n-drop/index.mjs";
import { Draggable as no } from "./Draggable.mjs";
import { Droppable as xo } from "./Droppable.mjs";
import { useDraggable as lo } from "./hooks/useDraggable.mjs";
import { useDroppable as co } from "./hooks/useDroppable.mjs";
import { kendoThemeMaps as Co } from "./theme.mjs";
import { memoizeOne as bo } from "./memoize.mjs";
import { Navigation as Eo } from "./navigation.mjs";
import { getActiveElement as Po, getInnerActiveElement as Ao } from "./getActiveElement.mjs";
import { TreeFieldsService as To } from "./tree-utils/FieldsService.mjs";
import { SortedPublicItemIds as vo } from "./tree-utils/SortedPublicItemIds.mjs";
import { addItem as yo, areAllDirectChildrenChecked as Lo, getAllDirectIndirectChildrenIds as Mo, getAllParents as wo, hasChildren as No, isEnabledAndAllParentsEnabled as ko, isItemExpandedAndWithChildren as Oo, removeItem as Uo, resolveItemId as Ro, resolveItemsIds as Wo, updateItem as _o } from "./tree-utils/itemUtils.mjs";
import { getNestedValue as Ho, isArray as Vo } from "./tree-utils/misc.mjs";
import { RowHeightService as zo } from "./rowHeightService.mjs";
import { UnstyledContext as Zo, useUnstyled as Xo } from "./unstyled/main.mjs";
import { uAnimation as qo } from "./unstyled/animations.mjs";
import { uSvgIcon as Qo } from "./unstyled/icons.mjs";
import { uButton as er, uButtonGroup as or, uDropDownButton as rr } from "./unstyled/buttons.mjs";
import { uComboBox as pr, uDropDownList as sr, uDropDownsActionSheet as ar, uDropDownsBase as mr } from "./unstyled/dropdowns.mjs";
import { uCalendar as ir, uDateInput as xr, uDateTimePicker as fr, uTime as lr, uTimePicker as ur } from "./unstyled/dateinputs.mjs";
import { radioButtonClasses as cr, radioGroupClasses as Ir, uInput as Cr, uMaskedTextBox as hr, uRadioButton as br, uRadioGroup as gr, uTextBox as Er } from "./unstyled/inputs.mjs";
import { uError as Pr, uFloatingLabel as Ar, uHint as Sr, uLabel as Tr } from "./unstyled/labels.mjs";
import { uForm as vr } from "./unstyled/form.mjs";
import { uPopup as yr } from "./unstyled/popup.mjs";
import { uGrid as Mr } from "./unstyled/grid.mjs";
import { actions as Nr, animationStyles as kr, base as Or, buttonPrefix as Ur, calendarPrefix as Rr, comboBoxPrefix as Wr, components as _r, containers as Gr, cssUtils as Hr, cursor as Vr, dateInputs as jr, ddbPrefix as zr, dimensions as Kr, directionMap as Zr, dropDownListPrefix as Xr, elements as Yr, fillModeMap as qr, forms as Jr, grid as Qr, gridPrefix as $r, icon as et, inputPrefix as ot, inputs as rt, jsonTheme as tt, labels as pt, maskedPrefix as st, orientationMap as at, pickerPrefix as mt, popup as nt, radioPrefix as it, roundedMap as xt, sizeMap as ft, states as lt, themeColorMap as ut } from "./unstyled/json-classes.mjs";
export {
  de as AsyncFocusBlur,
  a as BrowserSupportService,
  ke as COLLECTION_ACTION,
  ao as DragAndDrop,
  no as Draggable,
  xo as Droppable,
  Y as FIELD_REGEX,
  q as FOCUSABLE_ELEMENTS,
  J as FOCUSABLE_ELEMENTS_BASE,
  L as FormComponent,
  Ke as Icon,
  qe as IconWrap,
  Qe as IconsContext,
  W as KEYS,
  _ as Keys,
  Eo as Navigation,
  zo as RowHeightService,
  vo as SortedPublicItemIds,
  Xe as SvgIcon,
  Q as TABBABLE_ELEMENTS,
  To as TreeFieldsService,
  po as Typography,
  Zo as UnstyledContext,
  xe as WatermarkOverlay,
  t as ZIndexContext,
  Nr as actions,
  yo as addItem,
  kr as animationStyles,
  E as applyTrappedFocus,
  Lo as areAllDirectChildrenChecked,
  Or as base,
  Ur as buttonPrefix,
  Rr as calendarPrefix,
  n as canUseDOM,
  x as canUseRef,
  l as classNames,
  c as clone,
  I as cloneArray,
  C as cloneDate,
  h as cloneObject,
  b as cloneValue,
  Wr as comboBoxPrefix,
  _r as components,
  Gr as containers,
  Ie as createPropsContext,
  Hr as cssUtils,
  Vr as cursor,
  jr as dateInputs,
  zr as ddbPrefix,
  Kr as dimensions,
  Zr as directionMap,
  D as disableNavigatableContainer,
  le as dispatchEvent,
  Xr as dropDownListPrefix,
  Yr as elements,
  P as enableNavigatableContainer,
  ee as extendDataItem,
  qr as fillModeMap,
  A as firstFocusableChild,
  S as focusFirstFocusableChild,
  T as focusLastFocusableChild,
  B as focusableChildren,
  Jr as forms,
  Po as getActiveElement,
  Mo as getAllDirectIndirectChildrenIds,
  wo as getAllParents,
  Ao as getInnerActiveElement,
  oe as getItemPath,
  Ho as getNestedValue,
  j as getScrollbarWidth,
  w as getTabIndex,
  k as getter,
  Qr as grid,
  $r as gridPrefix,
  U as guid,
  No as hasChildren,
  ne as hasRelativeStackingContext,
  et as icon,
  ot as inputPrefix,
  rt as inputs,
  Vo as isArray,
  ko as isEnabledAndAllParentsEnabled,
  Oo as isItemExpandedAndWithChildren,
  tt as jsonTheme,
  v as keepFocusInContainer,
  Co as kendoThemeMaps,
  pt as labels,
  F as lastFocusableChild,
  re as mapTree,
  te as mapTreeItem,
  st as maskedPrefix,
  bo as memoizeOne,
  H as noop,
  at as orientationMap,
  mt as pickerPrefix,
  nt as popup,
  cr as radioButtonClasses,
  Ir as radioGroupClasses,
  it as radioPrefix,
  Uo as removeItem,
  Ro as resolveItemId,
  Wo as resolveItemsIds,
  xt as roundedMap,
  z as setScrollbarWidth,
  Z as setter,
  se as shouldShowValidationUI,
  ft as sizeMap,
  lt as states,
  u as strippedClassNames,
  eo as svgIconPropType,
  ut as themeColorMap,
  oo as toIconClass,
  ro as toIconName,
  e as treeIdUtils,
  qo as uAnimation,
  er as uButton,
  or as uButtonGroup,
  ir as uCalendar,
  pr as uComboBox,
  xr as uDateInput,
  fr as uDateTimePicker,
  rr as uDropDownButton,
  sr as uDropDownList,
  ar as uDropDownsActionSheet,
  mr as uDropDownsBase,
  Pr as uError,
  Ar as uFloatingLabel,
  vr as uForm,
  Mr as uGrid,
  Sr as uHint,
  Cr as uInput,
  Tr as uLabel,
  hr as uMaskedTextBox,
  yr as uPopup,
  br as uRadioButton,
  gr as uRadioGroup,
  Qo as uSvgIcon,
  Er as uTextBox,
  lr as uTime,
  ur as uTimePicker,
  _o as updateItem,
  Se as useAsyncFocusBlur,
  Oe as useCollection,
  Le as useCustomComponent,
  Pe as useDir,
  Re as useDocument,
  lo as useDraggable,
  co as useDroppable,
  He as useId,
  je as useIsomorphicLayoutEffect,
  Fe as useMouse,
  we as usePropsContext,
  Be as useRtl,
  Xo as useUnstyled,
  _e as useWindow,
  p as useZIndexContext,
  ae as validatePackage,
  be as withIdHOC,
  Ce as withPropsContext,
  Ee as withUnstyledHOC
};
